torch/csrc/jit/codegen/cuda/ir_graphviz.cpp

```diff
@@ -0,0 +1,451 @@
+}
+
+void IrGraphGenerator::printValue(const Val* val, const std::string& label) {
+  graph_def_ << "    " << getid(val) << " [label=\"" << label
+             << "\", shape=rect, color=green, fontsize=10];\n";
+}
+
+std::string IrGraphGenerator::generate() {
+  // IrGraphGenerator instances are not reusable
+  TORCH_CHECK(graph_def_.str().empty());
+  TORCH_CHECK(visited_.empty());
+
+  // record detail level
+  graph_def_ << "// detail level: ";
+  switch (detail_level_) {
+    case DetailLevel::ComputeOnly:
+      graph_def_ << "compute only\n";
+      break;
+    case DetailLevel::Basic:
+      graph_def_ << "minimal\n";
+      break;
+    case DetailLevel::Explicit:
+      graph_def_ << "explicit\n";
+      break;
+    case DetailLevel::Verbose:
+      graph_def_ << "verbose\n";
+      break;
+    default:
+      TORCH_CHECK(!"Unexpected detail level");
+  }
+
+  graph_def_ << "digraph fusion_ir {\n"
+             << "  node [shape=circle, color=gray];\n"
+             << "  edge [color=black];\n";
+
+  // Compute graph
+  generateComputeGraph();
+
+  // Schedule graph
+  if (detail_level_ > DetailLevel::ComputeOnly) {
+    generateScheduleGraph();
+  }
+
+  // All expressions & values
+  // (These are otherwise unreacheable (dead) nodes)
+  if (detail_level_ >= DetailLevel::Verbose) {
+    for (const auto* expr : fusion_->unordered_exprs()) {
+      handle(expr);
+    }
+    for (const auto* val : fusion_->vals()) {
+      handle(val);
+    }
+  }
+
+  // Finally, print all arc definitions
+  for (const auto& arc : arcs_) {
+    graph_def_ << "  " << arc << ";\n";
+  }
+
+  graph_def_ << "}\n";
+
+  // Make sure that all referenced nodes have been visited
+  for (const auto& kv : id_map_) {
+    TORCH_CHECK(visited(kv.first));
+  }
+
+  return graph_def_.str();
+}
+
+void IrGraphGenerator::generateComputeGraph() {
+  graph_def_ << "  subgraph cluster_compute {\n"
+             << "    label=\"compute\";\n"
+             << "    style=dashed;\n";
+
+  // Inputs
+  for (const auto* input : fusion_->inputs()) {
+    handle(input);
+  }
+
+  // Outputs
+  for (const auto* output : fusion_->outputs()) {
+    handle(output);
+  }
+
+  graph_def_ << "  }\n";
+}
+
+void IrGraphGenerator::generateScheduleGraph() {
+  graph_def_ << "  subgraph cluster_schedule {\n"
+             << "    label=\"schedule\";\n"
+             << "    style=dashed;\n";
+
+  // Connect TensorView with their TensorDomain
+  // (this will trigger the traversal of the schedule graph)
+  for (auto tv : tensor_views_) {
+    addArc(tv->domain(), tv, "[style=dashed, arrowhead=none]");
+    if (detail_level_ >= DetailLevel::Explicit) {
+      const auto root_domain = tv->getRootDomain();
+      addArc(tv, root_domain, "[style=dashed, color=green, arrowhead=none]");
+    }
+  }
+
+  graph_def_ << "  }\n";
+}
+
+void IrGraphGenerator::handle(const Statement* s) {
+  OptInConstDispatch::handle(s);
+};
+
+void IrGraphGenerator::handle(const Val* v) {
```

Comment:
nitpick.
I think if we use the Dispatch thing, you would want to go through `handle(Statement/Val/Expr)` via definition in inherited-scope and optin/out handles for the rest of the nodes.
But I see how it could save code in your implemented, so I'm totally fine with it if you decide to leave it as is.

However, if we are re-implementing all these stock functions from the inherited class, I don't see much value in inheritance here, especially that we now have branch to skip the call to inherited handle function.

Reply:
> via definition in inherited-scope and optin/out handles for the rest of the nodes",
I'm not sure I follow, can you please elaborate?

> especially that we now have branch to skip the call to inherited handle function.
Maybe I didn't parse this right, which "skip the call" are you talking about?

Reply:
Sorry for poor English :bow: 
> via definition in inherited-scope and optin/out handles for the rest of the nodes
This is basically what I was suggesting in the other comment, take the big hammer of `using OptInConstDispatch::handle` and get rid of the few `handle(Statement/Val/Expr)`.

> skip the call to inherited handle function.
I'm talking about the if statement in the function. We only calls `OptInConstDispatch::handle` when current node has not been `visited`. This is different from how inherited class traverses nodes.
https://github.com/csarofeen/pytorch/pull/44/files/4d96c2c534a641ab161d3f3bfb958e0848853b7c#diff-83aa246d49f3b1bca5f60e92bcaf616bR291
Basically inherited class only serves the purpose of an abstract interface. I guess I still see value here, as they would raise error when new nodes are added to codebase but not to `IrGraphGenerator`.

Reply:
As you noted, handing Val/Expr are not pass-through to the base implementation, so using would only save us one override: handle(Statement*)

> Basically inherited class only serves the purpose of an abstract interface. I guess I still see value here, as they would raise error when new nodes are added to codebase but not to IrGraphGenerator.

Sorry, I'm still not sure I understand: IrGraphGenerator implements the OptInConstDispatch for dispatching (which is a hidden part of its implementation, thus the private inheritance).
How would we do the dispatch w/o inheritance?